libs/common/RespReadUtils.cs

```diff
@@ -4,156 +4,363 @@
+        /// and, if successful, moves the given ptr to the end of the length header.
         /// </summary>
-        public static bool Read64Int(out long number, ref byte* ptr, byte* end)
+        /// <param name="len">If parsing was successful, contains the extracted length from the header.</param>
+        /// <param name="ptr">The starting position in the RESP string. Will be advanced if parsing is successful.</param>
+        /// <param name="end">The current end of the RESP string.</param>
+        /// <param name="isArray">Whether to parse an array length header ('*...\r\n') or a string length header ('$...\r\n').</param>
+        /// <returns>True if a length header was successfully read.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ReadLengthHeader(out int len, ref byte* ptr, byte* end, bool isArray = false)
         {
-            number = 0;
-            if (ptr + 3 >= end)
+            len = -1;
+            if (ptr + 3 > end)
                 return false;
 
-            Debug.Assert(*ptr == ':');
+            var readHead = ptr + 1;
+            var negative = *readHead == '-';
 
-            ptr++;
-            number = *ptr++ - '0';
-            while (*ptr != '\r')
+            // String length headers must start with a '$', array headers with '*'
+            if (*ptr != (isArray ? '*' : '$'))
             {
-                Debug.Assert(*ptr >= '0' && *ptr <= '9');
-                number = number * 10 + *ptr++ - '0';
-                if (ptr >= end)
+                RespParsingException.ThrowUnexpectedToken(*ptr);
+            }
+
+            // Special case: '$-1' (NULL value)
+            if (negative)
+            {
+                if (readHead + 4 > end)
+                {
                     return false;
+                }
+                if (*(uint*)readHead == MemoryMarshal.Read<uint>("-1\r\n"u8))
+                {
+                    ptr = readHead + 4;
+                    return true;
+                }
+                readHead++;
             }
-            ptr += 2;
+
+            // Parse length
+            if (!TryReadUlong(ref readHead, end, out var value, out var bytesRead))
+            {
+                return false;
+            }
+
+            if (bytesRead == 0)
+            {
+                RespParsingException.ThrowUnexpectedToken(*ptr);
+            }
+
+            // Validate length
+            len = (int)value;
+
+            if (negative)
+            {
+                RespParsingException.ThrowInvalidStringLength(-len);
+            }
+
+            if (value > int.MaxValue)
+            {
+                RespParsingException.ThrowIntegerOverflow();
+            }
+
+            // Ensure terminator has been received
+            ptr = readHead + 2;
             if (ptr > end)
+            {
                 return false;
+            }
+
+            if (*(ushort*)readHead != MemoryMarshal.Read<ushort>("\r\n"u8))
```

Comment:
I wonder if this can be JIT-optimized to avoid repeated parse; perhaps

``` c#
internal static readonly ushort CrLf = BitConverter.IsLittleEndian ? (ushort)0x0A0D : (ushort)0x0D0A; // see: ASCII
```

which *should* be treated like `const` by modern JITs

Reply:
``MemoryMarshal.Read<ushort>("\r\n"u8)`` is folded from as constant from the RVA static too. Either will work.

Reply:
> edit: actually static readonly fields can't be considered const.

Yes they can - see https://github.com/dotnet/runtime/issues/4287 - this is why reflection against static readonly fields was hard blocked at (some version, when the JIT change was added)

Reply:
Interesting.. Here's `FullOpts` for:
```CSharp

internal unsafe static class Resp
{
    internal static readonly ushort CrLf = BitConverter.IsLittleEndian ? (ushort)0x0A0D : (ushort)0x0D0A; 

    public static bool EqNewline_StaticGet(ref byte* ptr) => *(ushort*)ptr != CrLf;
}
```

```asm
; Assembly listing for method DisasmoPlayground.Resp:EqNewline_StaticGet(byref):ubyte (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,  3   )   byref  ->  rcx         single-def
;  V01 OutArgs      [V01    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V02 rat0         [V02,T01] (  2,  4   )     int  ->  rbx         "Spilling to split statement for tree"
;
; Lcl frame size = 32

G_M48716_IG01:  ;; offset=0x0000
       push     rbx
       sub      rsp, 32
						;; size=5 bbWeight=1 PerfScore 1.25
G_M48716_IG02:  ;; offset=0x0005
       mov      rax, qword ptr [rcx]
       movzx    rbx, word  ptr [rax]
       test     byte  ptr [(reloc 0x7ffd7b16899a)], 1      ; global ptr
       je       SHORT G_M48716_IG05
						;; size=15 bbWeight=1 PerfScore 8.00
G_M48716_IG03:  ;; offset=0x0014
       movzx    rax, word  ptr [(reloc 0x7ffd7b1689a0)]      ; static handle
       cmp      ebx, eax
       setne    al
       movzx    rax, al
						;; size=15 bbWeight=1 PerfScore 3.50
G_M48716_IG04:  ;; offset=0x0023
       add      rsp, 32
       pop      rbx
       ret      
						;; size=6 bbWeight=1 PerfScore 1.75
G_M48716_IG05:  ;; offset=0x0029
       mov      rcx, 0x7FFD7B168968
       mov      edx, 2
       call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       jmp      SHORT G_M48716_IG03
						;; size=22 bbWeight=0 PerfScore 0.00

; Total bytes of code 63, prolog size 5, PerfScore 14.50, instruction count 17, allocated bytes for code 63 (MethodHash=ce0a41b3) for method DisasmoPlayground.Resp:EqNewline_StaticGet(byref):ubyte (FullOpts)
; ============================================================
```

Reply:
This might actually be just be blamed on Disasmo. There might depend on timing of the static ctor..

Reply:
> on net8, all three approaches are basically identical (within measurability); the only delta is on down-level TFMs without modern JIT - on net48, for example, static readonly fields are the clear winner, and memorymarshal read is the clear loser: https://gist.github.com/mgravell/0f9306681f3ae6bbff2ee4891783f303

Ah, thanks for benchmarks! Seems like the rule to always measure is right in different way this time 😄  _Hopefully_ Garnet will never ship parts of it to legacy .NET.

Reply:
We have recently been converting large swaths of Garnet protocol code to use `MemoryMarshal.Read` which improves code readability compared to the random number comparisons. We are only supporting .NET 6 and 8 for now, but this result means that if we ever support older .NET then we would need to revisit this choice.

Reply:
This is a super interesting discussion. Using `MemoryMarshal.Read` definitely results in much more readable and maintainable code, so I would suggest we keep using it for RESP parsing in the main parser loop.

We could still hard-code common checks such as `CrLf` or wrap those checks in separate inline functions, so that we could easily change the implementation in the future, if we believe we'll ever need to support older .net versions.

Reply:
I'm still with @mgravell here and really think comparing against a field called ``CrLf`` or ``Newline`` is more readable than MM.Read + RVA magic that is more expensive for JIT and potentially eats away from its tight budget to do other optimizations. But the difference is so miniscule so either is fine.

Reply:
Agreed on this. Shouldn't waste the JIT budget unless we get good returns.

Reply:
I am resolving this conversation for now. We should consider this issue in the next cleanup for the parsing logic. In particular, for readability it would be nice to just have a function of the form `Expect(NewLine)` (or similar) that does the check using a static field and handles any potential errors.